go/glide/lib.h

```diff
@@ -0,0 +1,59 @@
+#include <stdarg.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdlib.h>
+
+enum ErrorType {
+  ClosingError = 0,
+  RequestError = 1,
+  TimeoutError = 2,
+  ExecAbortError = 3,
+  ConnectionError = 4,
+};
+typedef uint32_t ErrorType;
+
+typedef struct RedisErrorFFI {
+  const char *message;
+  ErrorType error_type;
+} RedisErrorFFI;
+
+typedef struct ConnectionResponse {
+  const void *conn_ptr;
+  const struct RedisErrorFFI *error;
+} ConnectionResponse;
+
+/**
+ * Success callback that is called when a Redis command succeeds.
+ */
+typedef void (*SuccessCallback)(uintptr_t channel_address, const char *message);
+
+/**
+ * Failure callback that is called when a Redis command fails.
+ */
+typedef void (*FailureCallback)(uintptr_t channel_address, const struct RedisErrorFFI *error);
```

Comment:
To ease marshalling it is better to have
```suggestion
typedef void (*FailureCallback)(uintptr_t channel_address, ErrorType error_type, const char *message);
```
Otherwise we have to deal with memory allocations/deallocations for that pointer.

Reply:
I've already implemented the proper functions for freeing this struct and the string within. We can do this, but it goes against what's in the design doc and we'll have to make sure Aaron's work is aligned with this API change as well. I think we should discuss this first with the rest of the team before changing it.

Reply:
It is ok to adjust the design noc now

Reply:
Is it? @aaron-congo are you aware of these changes?

Reply:
I wasn't aware. We need the RedisErrorFFI and a mechanism to free it for the connection request, since we are not using a callback and C functions can only return a single result. If we need these things anyways, should we just use it in the failure callback as planned?